xenonpy/contrib/ismd/reactant_pool.py

```diff
@@ -0,0 +1,157 @@
+        """
+        A module consists the reactant pool for proposal
+        ----------
+        Parameters:
+            pool_df : a dataframe contains the all usable reactants,
+            sim_df : a matrix of similarity between each pair of reactant in pool_df, size len(pool_df)*len(pool_df).
+            reactor : reaction prediction model.
+            pool_smiles_col : the column name of reactant SMILES in the pool_df,
+            sim_id_in_pool : the column name of id in pool_df corresponding to sim_df, if None, use index.
+            sample_reactant_idx_col : the column name of reactant id in sample dataframe.
+            #sample_reactant_idx_old_col : the column name of old reactant id in sample dataframe, copied from sample_reactant_idx_col before proposal.
+            sample_reactant_smiles_col : the column name of reactant SMILES in sample dataframe
+            sample_product_smiles_col : the column name of product SMILES in sample dataframe
+            splitter : string used for concatenating reactant in a reactant set.
+        """
+        if len(sim_df) != len(sim_df.columns):
+            raise NotSquareError(len(sim_df), len(sim_df.columns))
+
+        if len(pool_df) != len(sim_df):
+            raise SimPoolnotmatchError(len(pool_df), len(sim_df))
+
+        if sim_id_in_pool is not None:
+            self._pool_df = pool_df.set_index(sim_id_in_pool)
+        else:
+            self._pool_df = pool_df
+        self._sim_df = sim_df
+        self._reactor = reactor
+        self._pool_smiles_col = pool_smiles_col
+        #self._sim_id_in_pool = sim_id_in_pool
+        self._sample_reactant_idx_col = sample_reactant_idx_col
+        #self._sample_reactant_idx_old_col = sample_reactant_idx_old_col
+        self._sample_reactant_smiles_col = sample_reactant_smiles_col
+        self._sample_product_smiles_col = sample_product_smiles_col
+        self._splitter = splitter
+
+    def single_index2reactant(self, reactant) -> str:
+        """
+        convert a list of index to string concatenated by the splitter.
+        ----------
+        Parameters:
+            reactant : list of reactant id (e.g. [45011, 29392])
+        Returns:
+            reactant_smiles : reactant SMILES concatenated by the splitter (e.g. CC(C)Br.COc1ccc(C=O)cc1O)
+        """
+        r_list = [self._pool_df.iloc[r][self._pool_smiles_col] for r in reactant]
+        #reactant_smiles = self._splitter.join(r_list)
+        return self._splitter.join(r_list)
+
+    def index2sim(self, r_idx_old):
+        """
+        substitute the reactant id by a similar one.
+        ----------
+        Parameters:
+            r_idx_old : reactant id to be substituted (e.g. 29392)
+        Returns:
+            r_idx_new : an id of reactant similar to r_idx_old (e.g. 9980)
+        """
+        if r_idx_old not in self._pool_df.index:
+            raise ReactantNotInPoolError(r_idx_old)
+        sim_col = self._sim_df[[r_idx_old]].drop(r_idx_old)
```

Comment:
What's the data type of `r_idx_old `?

if `r_idx_old ` is int or str, `self._sim_df[[r_idx_old]]` will return a series, then `.drop(r_idx_old)` will drop all data which have `label == r_idx_old `. So the next line confused me, because `sim_col ` has no date have with label equal to `r_idx_old `.

Reply:
basically, self._sim_df is a dataframe whose index and column name is the same, self._sim_df[[r_idx_old]] will return the column of named r_idx_old, .drop(r_idx_old) will remove the r_idx_old th element of this column. the next line: sim_col[r_idx_old] return the column itself.
I tested this code, it works, however, might be possible to rewrite it in a better way.

Reply:
OK, I see